subprojects/core/src/main/java/org/gradle/process/internal/ExecHandleRunner.java

```diff
@@ -79,6 +80,7 @@ public void run() {
                 streamsHandler.stop();
                 detached();
             } else {
+                CurrentBuildOperationRef.instance().clear();
```

Comment:
Is this the right spot?  Should this be happening if an exception is thrown from `streamsHandler.start` or `execHandle.started` or `startProcess`?

What does clearing the current thread's build operation do if we're then waiting for the process to stop?

Reply:
In those other cases, we're exiting the thread immediately, so a reference won't be held. Only in this case, where we are waiting on the process, does this thread hang around for longer than the build and hold references we don't need. Arguably we could clear it earlier but this is the only spot it matters, every other path exits quickly.

Reply:
Do this impact task actions in general? e.g., something like...
```
@TaskAction
void doIt() {
    // CurrentBuildOperationRef is non-null
    execOperations.exec {
        // configure exec
    }
    // CurrentBuildOperationRef is null?
}
```

Reply:
`ExecHandleRunner` is always used in its own thread (see `DefaultExecHandle#start`), though I guess I could make it preserve to be safe?

Reply:
This does give me another concern though, perhaps we should only be clearing the ExecHandleRunner's reference for workers. If we clear it regardless, we lose the operation context that an exec operation might otherwise have.